functionality review

oscamaina opened the pull request.

oscamaina committed:
Adds more tests

oscamaina committed:
initializes lists

oscamaina committed:
adds requirements list

oscamaina committed:
removed unnecessary files

oscamaina committed:
updated the requirement list

oscamaina committed:
refactor tests

oscamaina committed:
initial functionality implementation

oscamaina committed:
refactor tests

oscamaina committed:
remodelled method create_room

remodelled create_room method to use rooms model

oscamaina committed:
initial commit for rooms model

oscamaina committed:
initial commit for person model

oscamaina committed:
adds app interface

oscamaina committed:
implemented adding and allocating persons

oscamaina committed:
refactor tests

oscamaina committed:
modified classes

added occupants variable to room class,added capacity to classes office and livingSpace

oscamaina committed:
modified create_room method

modiefied create_room method to enable creating as many
rooms as possible by specifying multiple room names
after the create_room command

oscamaina committed:
refactor tests

fixed bug in method create_room to take a list of room names instead of
strings

oscamaina committed:
refactor tests

refactored to test the for inheritance

oscamaina committed:
modified functionalities

modified the functions to make use of the defined models

oscamaina committed:
moddified usage

updated the commands, modified command add_person to take first_name and
last_name arguments

oscamaina committed:
adds rooms model

oscamaina committed:
adds person model

oscamaina committed:
remove person.py

oscamaina committed:
remove test.py

Gchesta commented on dojo.py:
```diff
@@ -18,93 +19,85 @@ def create_room(self, room_type, room_names):
-						room = LivingSpace(room_name)
-						self.livingSpaces.append({'room_name': room_name, 'occupants': occupants, 'room_capacity': room_capacity})
-						msg += "Living space " + room_name + " added successfully \n"
+					room = Room(room_name)
```
Room is an abstract class. This means that it cannot be made on its on. For example, when God creates a human being, he creates a a male or female. We can't have just a person. Same thing with room. A room must either be LivingSpace or Office. It can't just be room.

Gchesta commented on dojo.py:
```diff
@@ -18,93 +19,85 @@ def create_room(self, room_type, room_names):
 				else:
 					msg += "Wrong room type \n"
-		return msg
+			return msg
```
Is the code working? The reason being, it seems to me that you're expecting a list. If a list has five items; what will happen to the four items once the first loop is completed and the function is terminated by a return command? 

I would also suggest that you make use of the print() function so that the user can see what's going on. I don't think a return statement will output something for the user

Gchesta commented on dojo.py:
```diff
@@ -18,93 +19,85 @@ def create_room(self, room_type, room_names):
-					msg = self.allocate_office(name)
-					return "Staff added successfully with " + str(id) + ". " + msg
+			id = len(self.all_people) + 1
+			person = Person(first_name, last_name)
```
Again, Person is an abstract class. You can only add a fellow or a staff. Not a person. Otherwise, "Andrew" in all_people will not be the  same as "Andrew" in fellows because one is an instance of Person the other is an instance of Fellow

Gchesta commented on dojo.py:
```diff
@@ -18,93 +19,85 @@ def create_room(self, room_type, room_names):
+			self.fellows.append(person)
+			if accomodation == 'Y':
+				return "Fellow added successfully with id " + str(id) + ". " \
+				+ self.allocate_office(full_name) + ". " + \
```
Great idea to implement a separate function to allocate space. However, what will happen if there are no available spaces? What message will be returned?

Gchesta commented on dojo.py:
```diff
@@ -18,93 +19,85 @@ def create_room(self, room_type, room_names):
+			else:
+				self.staffs.append(person)
+				return "Staff added successfully with id " + str(id) + ". " \
+				+ self.allocate_office(full_name)
```
Same as above

kevintumbo commented on test.py:
```diff
@@ -1,24 +0,0 @@
-        """Tests if room already exists"""
-        room =self.dojo.create_room("office", "Bujumbura")
-        self.assertEqual(room,"Room already exists")
-
```
when you say room already exists ....which room are you refering to?

oscamaina replied:
Room already exists acts as my second argument,

kevintumbo replied:
please elaborate further.

oscamaina replied:
How best can i put it?? in my method create_room, i first test if the room name already exists, in the case it exists, it returns a message Room already exits. From that am asserting the two, the message from 'room' and 'Room already exits'

kevintumbo replied:
ok. But from the test i see the only rooms that exist are A04 and Arusha. Won't Bujumbura be added?

kevintumbo commented on test.py:
```diff
@@ -1,24 +0,0 @@
-        """Tests if room already exists"""
-        room =self.dojo.create_room("office", "Bujumbura")
-        self.assertEqual(room,"Room already exists")
-
```
Oh ... ive seen these are changes made. my bad.

oscamaina committed:
updated file

updated the import lines, modified the methods allocate_office and
allocate_living to filter rooms with available spaces

oscamaina committed:
update return message

mnjaggah commented on dojo.py:
```diff
@@ -1,12 +1,103 @@
+		if category in ("fellow", "staff"):
+
+			id = len(self.all_people) + 1
+			person = Person(first_name, last_name)
```
from your person model, Person class accepts two values, name and category. Here you're passing the first name and last name to an instance of this class

oscamaina replied:
thanks, I need to push the updated person model

oscamaina committed:
renamed models to app

oscamaina committed:
deleted models directory

oscamaina committed:
adds room model

oscamaina committed:
adds person model

oscamaina committed:
moved dojo.py to app

oscamaina committed:
modified import statement

oscamaina committed:
modified import statement

oscamaina committed:
updated file dojo

modified the import statements and fixed the error of parsing 3
arguments in Person modef

ianoti commented on app.py:
```diff
@@ -0,0 +1,108 @@
+        + ' (type help for a list of commands.)'
+    print(__doc__)
+    prompt = '>> '
+    file = None
```
what is the purpose of this variable? why do you have it in your code?

ianoti replied:
Consider removing this line of code `file = None` it serves no purpose.

ianoti commented on app.py:
```diff
@@ -0,0 +1,108 @@
+    prompt = '>> '
+    file = None
+
+    funt = dojo.Dojo()
```
you already imported Dojo class why include the `dojo.Dojo()` consider instantiating as `funt` = `Dojo()` also improve variable naming, this isn't descriptive enough.

ianoti commented on app.py:
```diff
@@ -0,0 +1,108 @@
+        """
+        Usage: print_room <room_name>
+        """
+        room_name = (arg['<room_name>'])
```
You have docopt calls for `print_room` before I've reviewed your tests for the `print_room` method. Did you skip ahead to task 1? I like the initiative but I need to see your tests first.

oscamaina replied:
am yet to actually implement that, just but a layout

ianoti replied:
okay

ianoti commented on app.py:
```diff
@@ -0,0 +1,108 @@
+if __name__ == '__main__':
+    try:
+        Allocation().cmdloop()
+    except KeyboardInterrupt:
```
good thinking to handle keyboard interrupts.

oscamaina committed:
adds __init__ file

ianoti commented on app/dojo.py:
```diff
@@ -0,0 +1,118 @@
+			if found:
+				msg += "Room with name " + room_name + " already exists \n"
+			else:
+				if room_type in ('office', 'living'):
```
consider removing lines 30-33 the rooms created should only be instances of the `Office` or `LivingSpace` classes. you can append the rooms to a single list of all rooms after appending them to the respective lists of `office` and `living_space`

ianoti commented on app/dojo.py:
```diff
@@ -0,0 +1,118 @@
+	def __init__(self):
+		self.all_rooms = []
+		self.offices = []
+		self.livingSpaces = []
```
consistency in naming your variables consider `self.living_spaces = []`

ianoti commented on app/dojo.py:
```diff
@@ -0,0 +1,118 @@
+
+	def add_person(self, first_name, last_name, category, accomodation="N"):
+		""" function to add a person """
+		full_name = first_name.capitalize() + " " + last_name.capitalize()
```
Consider removing the `full_name` variable. Using this for printing is wasteful of memory space since you will have the object and can get the names from the object instead of from the arguments passed to the function.

ianoti commented on app/dojo.py:
```diff
@@ -0,0 +1,118 @@
+		category = category.lower()
+		accomodation = accomodation.upper()
+
+		if category in ("fellow", "staff"):
```
consider removing lines `52-56` the only classes you instantiate should be the `Fellow` and `Staff` classes.

ianoti commented on app/dojo.py:
```diff
@@ -0,0 +1,118 @@
+
+		if category in ("fellow", "staff"):
+
+			id = len(self.all_people) + 1
```
You have not tied the `id` field to the object. It should be an attribute of `Fellows` or `Staff` object.

oscamaina committed:
removed logical error

ianoti commented on app/dojo.py:
```diff
@@ -0,0 +1,118 @@
+		if len(living_with_space) > 0:
+			random_living_space = choice(living_with_space)
+			if len(random_living_space.occupants) < random_living_space.max_capacity:
+				random_living_space.occupants.append(full_name)
```
this is not the best way of storing occupants in the occupants list, consider the impact of having two people of the same name in a room? how would the system differentiate between them for reallocation, consider appending the fellow or staff object to this list.

ianoti commented on app/person.py:
```diff
@@ -0,0 +1,22 @@
+class Person(object):
+    """ class person """
+    def __init__(self, first_name, last_name, category = 'none'):
+        """ initializing person attributes """
```
you should add an `id` field for unique identification to the class.

ianoti commented on requirements.txt:
```diff
@@ -0,0 +1,5 @@
+appdirs==1.4.2
```
update your `requirements.txt` file.

ianoti commented on tests/test.py:
```diff
@@ -0,0 +1,82 @@
+
+        """
+        current_no_fellows = len(self.dojo.fellows)
+        self.dojo.add_person('Kipyegon Ken', 'Fellow')
```
your `add_person` method doesn't reflect the one you are testing here, refactor your tests if you adjust the functionality of the code.

ianoti commented on tests/test.py:
```diff
@@ -0,0 +1,82 @@
+        current_no_livingSpaces = len(self.dojo.livingSpaces)
+        living_space =self.dojo.create_room("living", ["Arusha"])
+        room = LivingSpace('java')
+        self.assertIsInstance(room, Room)
```
you should assertIsInstance of livingSpace not Room

ianoti commented on tests/test.py:
```diff
@@ -0,0 +1,82 @@
+        current_no_offices = len(self.dojo.offices)
+        office = self.dojo.create_room("office", ["A04"])
+        room = Office('kigali')
+        self.assertIsInstance(room, Room)
```
you should assert is instance of Office not Room

ianoti reviewed (commented):
Good job on the Task. Implement the changes I have suggested and you should be done with the tasks for Task 0 and can proceed to make a video.

## Suggestions
Include code to handle bad user input like giving numbers as a name for people in the system or rooms, the names should all be validated to check if they are strings before execution.

oscamaina committed:
refactor tests

oscamaina committed:
updates file

oscamaina committed:
updates file

removed unnecessary code, updated the allocate function to append either
intance of fellow or staff

oscamaina committed:
adds attribute id

oscamaina committed:
updates the file

oscamaina committed:
fixed error of maximum capacity

oscamaina committed:
Fixed bug of ModuleNotFound

oscamaina committed:
removed logical error

oscamaina committed:
corrects return message

ianoti commented on app/dojo.py:
```diff
@@ -0,0 +1,114 @@
+
+		if len(office_with_space) > 0:
+			random_office = choice(office_with_space)
+			if len(random_office.occupants) < random_office.max_capacity:
```
its unnecessary to handle this possibility in lines 88. The offices in `office_with_space` all have space available.

oscamaina replied:
thanks for that identification

ianoti commented on app/dojo.py:
```diff
@@ -0,0 +1,114 @@
+
+		living_with_space = []
+		for living_allocate in self.living_spaces:
+			if len(living_allocate.occupants) < living_allocate.max_capacity:
```
line 102 is unnecessary, the `living_spaces` all have space.

ianoti reviewed (commented):
Make the minor changes, send me the video, and you'll have the go-ahead to move to Task 1.

oscamaina committed:
removed unnecessary line

oscamaina committed:
removed unnecessary code lines and added comments

oscamaina committed:
added unittests

added unittests for print_room, print_allocations and print_unallocated

oscamaina committed:
adds class methods

added methods print_room, print_allocations and print_unallocated
